Fixed #36134 -- Add support for ABSENT ON NULL clause for JSONField.

john-parton opened the pull request:
#### Trac ticket number

https://code.djangoproject.com/ticket/36134

#### Branch description
On all backends except Sqlite and Postgres version 15 and earlier, there is an additional clause to the JSON_ARRAY SQL which controls whether SQL NULL values are mapped to JSON null values or whether they are omitted.

In sqlite and postgres 15 and earlier, the default behavior is to map them, so my initial feature for JsonArray just had that as the only behavior without any ability to customize on the richer backends.

This adds a parameter to JsonArray, defaulting to the old behavior.

#### Checklist
- [ ] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [ ] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [ ] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

john-parton commented:
A few notes real quick:

I chose to use an enum rather than a simple boolean keyword argument to mirror the SQL

I put the enum on the JsonArray class, because I didn't think it had any use outside of JSON_ARRAY. I actually am seeing that's not true. It looks like JsonObject has a similar clause: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/JSON_OBJECT.html I could expand this pull request to add the functionality to JsonObject as well. I don't think it would be particularly difficult.

Docs were a pretty rough copy/paste/tweak job.

I added a new DB feature. I originally just did some basic checks inside the class, but it made it basically impossible to test cleanly.

charettes commented on django/db/models/functions/json.py:
```diff
@@ -9,25 +11,40 @@ class JSONArray(Func):
+        NULL = "NULL ON NULL"
+        ABSENT = "ABSENT ON NULL"
+
+    def __init__(self, *expressions, on_null=OnNull.NULL):
```
We've opted to use boolean flags instead for these kind of options (e.g. see `OrderBy(nulls_last)`, UniqueConstraint(nulls_distinct)` and others).

```suggestion
    def __init__(self, *expressions, absent_on_null=False):
```

Not sure about what the kwarg should be named though. It doesn't have to be a 1:1 map with the SQL feature name so maybe `exclude_nulls=False` could do?

john-parton replied:
Happy to use a bool if you think that's best. I sort of start with SQL and work my way backwards.

john-parton replied:
I've got a top-level comment, but I'm not sure the best way to proceed if we want to also include JsonObject. I think it probably would be best to try and include JsonObject, especially because I don't want to introduce TWO features or repurpose a feature that explicitly has the word `array` in it.

john-parton commented:
It looks like the biggest issue with adding an `on_null` clause to `JsonObject` is that it already accepts variadic `**kwargs`, so in the case that you want to make a json object that has an `on_null` _key_, you wouldn't be able to do it.

The closest parallel I can think of in the existing database is with `get_or_create` where if you want to do a lookup on a field called `defaults`, you must specify `defaults__exact`

john-parton commented:
As an aside, my version of JSONObject I made long ago didn't take `**kwargs`, but rather you were supposed to pass everything as arguments

So instead of

```python
JSONObject(
    key="value"
)
```

You would do
```
JSONObject(
    Value("key"), F("value")
)
```

Which is obviously more verbose, but it let you create an object where one of the keys was actually dynamic from another column.

It would also resolve the ambiguity of providing an `absent_on_null` flag, because it didn't take any keyword arguments.

I'm not exactly sure how that would resolve the issue with current django, though.

Another thought would be to have an alternate constructor as a classmethod

```
JSONObject.absent_on_null(**kwargs)
```

Or having to manually set a prop.

```
obj = JSONObject(...)
obj.absent_on_null = True
```

But the prop is error-prone and requires you to create a temporary variable.

charettes commented:
I'd focus this PR on `JSONArray` and see what can be done for `JSONObject` in a distinct ticket/MR?

A potential solution could be to require that fields are provided as a `list[str, Resolvable]` when nulls exclusion is requested so allowed signatures would be of the form

```python
@overload
def __init__(self, fields: list[str, Resolvable], *, exclude_nulls: bool = False): ...

@overload
def __init__(self, **fields: Resolvable): ...

def __init__(self, *args, **kwargs):
    if len(args) > 1:
        raise TypeError
    elif args:
        fields = args
        exclude_nulls = kwargs.pop("exclude_nulls", False)
        if kwargs:
            raise TypeError
    else:
        fields = kwargs
```

charettes commented on django/db/models/functions/json.py:
```diff
@@ -9,25 +9,36 @@ class JSONArray(Func):
             raise NotSupportedError(
                 "JSONFields are not supported on this database backend."
             )
+        if self.absent_on_null and not connection.features.supports_json_absent_on_null:
+            raise NotSupportedError(
+                "ABSENT ON NULL is not supported by this database backend."
+            )
```
Could this potentially be emulated on the Python side by having a converted that simply that removes `None` members?

```python
def from_db_value(self, value):
    return [v for v in value if value is not None]
```

john-parton replied:
Hmm... That's a great point. Would that work with `.values()` and other less common use cases?

john-parton replied:
I'll try adding it and removing the settings completely, then run tests.

john-parton replied:
Actually, I don't know how `from_db_value` works in this context. I thought that was more used on fields, and not on expressions.

charettes replied:
> I don't know how from_db_value works in this context. I thought that was more used on fields, and not on expressions.

You could assign a specialized `JSONField` subclass instance with `from_db` like the above.

john-parton replied:
Yeah, I think it's possibly a foot gun.

john-parton replied:
Postgres 15 is EoL in 2027, which isn't terribly far away. I don't know if SQLite will ever introduce the clause, but I think SQLite users are generally used to quirkiness.

john-parton replied:
There is a way to emulate it, but you're not going to like it. Basically just wrap every item up so that if it's null, it's an empty array, otherwise an array containing just the element, and then concatenate all the arrays.

This IS possible on postgres, but I have no idea how to do it on sqllite


```sql
JSON_ARRAY(<expression1>, <expression2>, ... ABSENT ON NULL)
```

Becomes

```sql
COALESCE(jsonb_build_array(<expression1>), []::jsonb) ||
COALESCE(jsonb_build_array(<expression2>), []::jsonb) ...
```

john-parton replied:
Coalesce didn't really work the way I wanted to for reasons I couldn't quite understand, but here's proof that you can do it using CASE/WHEN:

Oh boy, it actually passes the basic tests.

```python3
def as_postgresql(self, compiler, connection, **extra_context):
    # Casting source expressions is only required using JSONB_BUILD_ARRAY
    # or when using JSON_ARRAY on PostgreSQL 16+ with server-side bindings.
    # This is done in all cases for consistency.
    casted_obj = self.copy()
    casted_obj.set_source_expressions(
        [
            (
                # Conditional Cast to avoid unnecessary wrapping.
                expression
                if isinstance(expression, Cast)
                else Cast(expression, expression.output_field)
            )
            for expression in casted_obj.get_source_expressions()
        ]
    )

    if connection.features.is_postgresql_16:
        return casted_obj.as_native(
            compiler, connection, returning="JSONB", **extra_context
        )
    
    num_expressions = len(casted_obj.get_source_expressions())

    if self.absent_on_null and num_expressions > 0:
        from django.db.models import Case, When
        from django.db.models.functions import Concat  # avoid circular import

        args = [
            Case(
                When(IsNull(expression, True), then=JSONArray()),
                default=JSONArray(expression),
            ) for expression in casted_obj.get_source_expressions()
        ]

        if len(args) > 1:            
            return Concat(*args).as_sql(
                compiler,
                connection,
                **extra_context,
            )
        else:
            return args[0].as_sql(
                compiler,
                connection,
                **extra_context,
            )
 ```

john-parton replied:
OK, so going down the rabbit hole even further, there's not even a standard way to get json length

Here's what I've come up with real quick

```python
class JSONArrayLength(Func):
    output_field = IntegerField()
    arity = 1

    def __init__(self, *expressions, absent_on_null=False):
        self.absent_on_null = absent_on_null
        super().__init__(*expressions)

    def as_sql(self, compiler, connection, **extra_context):
        if not connection.features.supports_json_field:
            raise NotSupportedError(
                "JSONFields are not supported on this database backend."
            )
        return super().as_sql(compiler, connection, **extra_context)
    
    def as_mysql(self, compiler, connection, **extra_context):
        return super().as_mysql(compiler, connection, function="JSON_LENGTH", **extra_context)
    
    def as_oracle(self, compiler, connection, **extra_context):
        return super().as_sql(
            compiler, 
            connection, 
            template="json_value(%(expressions)s, '$.size()')", 
            **extra_context,
        )
    
    def as_postgresql(self, compiler, connection, **extra_context):
        return super().as_sql(compiler, connection, function="JSONB_ARRAY_LENGTH", **extra_context)
    
    def as_sqlite(self, compiler, connection, **extra_context):
        return super().as_sql(compiler, connection, function="JSON_ARRAY_LENGTH", **extra_context)
 ```
 
 The oracle one is weird because there isn't a dedicated function.

john-parton replied:
I went ahead and pushed the workaround code for MySQL and Postgres 15 and earlier, so we can decide how bad it is.

john-parton commented:
> I'd focus this PR on `JSONArray` and see what can be done for `JSONObject` in a distinct ticket/PR?
> 
> A potential solution could be to require that fields are provided as a `list[str, Resolvable]` when nulls exclusion is requested so allowed signatures would be of the form
> 
> ```python
> @overload
> def __init__(self, fields: list[str, Resolvable], *, exclude_nulls: bool = False): ...
> 
> @overload
> def __init__(self, **fields: Resolvable): ...
> 
> def __init__(self, *args, **kwargs):
>     if len(args) > 1:
>         raise TypeError
>     elif args:
>         fields = dict(args)
>         exclude_nulls = kwargs.pop("exclude_nulls", False)
>         if kwargs:
>             raise TypeError
>     else:
>         fields = kwargs
> ```
> 
> It also seems like a less useful feature for `JSONObject` than it is for `JSONArray` or even `JSONArrayAgg` IMO.

Ack, I wish the type annotations were just in django. That's so clear.

It's funny, I was hacking on pretty much exactly this solution , minus type annotations.

john-parton commented on django/db/models/functions/json.py:
```diff
@@ -9,25 +9,36 @@ class JSONArray(Func):
             connection,
             template=(
-                f"%(function)s(%(expressions)s {null_on_null} RETURNING {returning})"
+                f"%(function)s(%(expressions)s {on_null_clause} RETURNING {returning})"
```
So this is a VERY minor quibble, but if you don't have any source expressions, the generate SQL will end up having a triple space in it.

If the spaces were moved to lines 33 and 35, and then removed from the template here, it would resolve that, at the expense of making it slightly awkward right here.

Is that even something worth discussing?

john-parton replied:
if I were hand-writing the SQL, I would write it like `JSON_OBJECT(RETURNING JSONB)`, but the ORM will emit `JSON_OBJECT(   RETURNING JSONB)`

john-parton commented:
I removed `array` from the DB feature and added a comment that it's supposed to indicate support for the clause in both array and object contexts.

john-parton commented:
Looks like MySQL tests are failing. I think it's because it doesn't use the "as_native" definition where the code is.

I'll take a closer look.

john-parton commented:
Right, because MySQL doesn't support it. Duh. Toggled the value on the backend. Need to update the docs.

I am wondering if a workaround for the non-native backends would be worth it. I know a lot of bending over backwards is done for some functions and backends.

john-parton commented:
Alright, thought about it for a bit and decided the workaround wasn't as nasty as I thought it might be, at least for MySQL and Postgres 15.

I have absolutely no idea how to do it on SQLite. I believe it would require a subquery and an aggregate, or something really, really weird. Probably not safe to do it automatically/transparently.

john-parton commented:
Alright, last suite of changes:

* Removed DB feature, because all the database backends except SQLite now have similar behavior
* Reworked tests for `If`/`Unless` to check for SQLite, instead of DB feature
* Updated docs to remove note that Postgres 15 doesn't work

john-parton commented:
Polishing:

* There was some imports inside a method to avoid a circular import, but they were a 'shortcut' import from a parent module. Explicitly importing from where classes were defined removed the need for that.
* There was a small duplication of a code in `as_postgres` and `as_mysql` to handle the case where there are no expressions. I moved it into the workaround method to keep it DRY, but it does result in some complicated `super()` invocations that might be difficult to reckon with. I believe there needs to be more test coverage for when expressions == [] and absent_on_null == True

john-parton commented:
Few extra tests added, tests reorganized a bit:

* Test absent_on_null with an empty array
* Test absent_on_null with two elements, one null and one non-null
* Test absent_on_null with one element, both the case of a null element and non-null element

I think that's should be sufficient to exercise all of the branches in the workaround.

john-parton commented on tests/db_functions/json/test_json_array.py:
```diff
@@ -107,6 +108,51 @@ def test_order_by_nested_key(self):
+            ).first()
+
+
+@unittest.skipIf(
+    connection.vendor == "sqlite",
+    "SQLite does not support ABSENT ON NULL.",
+)
```
I assume you mean 3rd party backends? Or backends we might add in the future?

Both?

charettes replied:
yes, third party backends heavily rely on feature flags to denote what they support and don't.

john-parton replied:
Makes sense. I don't want to bikeshed to much on what the specific feature flag should be called. I was trying to avoid that discussion completely, tbh.

Two candidates:

* `supports_json_absent_on_null`
* `supports_json_array_absent_on_null`

charettes replied:
Do any backends support `ABSENT ON NULL` on `JSON_ARRAY` but not other JSON functions such as `JSON_ARRAYAGG`, and `JSON_OBJECT`? If not `supports_json_absent_on_null` seems like the best candidate.

john-parton replied:
I think it would depend on whether a workaround is _possible_.

Without exhaustively going through all the database backends, it might not be possible to sanely implement the functionality using whatever database-specific concatenation features are available.

I think this is going to be vary in difficulty depending on the backend.

john-parton replied:
So for instance, if a workaround isn't possible for JSON_OBJECT using MySQL, and we just have one shared feature flag, then we'll be in a bit of a mismatched situation if we want to implement `absent_on_null` for JsonObject.

charettes replied:
That would be to backends to figure out then and request distinct `supports_json_[type]_absent_on_null` feature flags if they run into this problem. `supports_json_absent_on_null` seems like a good start.

john-parton replied:
That's a sane position, I think. I'll refactor.

john-parton commented:
Additional changes:

* Restored the feature flag that an earlier version had. 3rd party backends will need to reach out if there's a backend that only has partial support for ABSENT ON NULL across the various functions.
* Merged the two test classes again. I realized the class I created didn't have any setup, so I just squashed them back together
* applied `isort`. I'll try to be more fastidious with this. I'm just used to my commit hooks and `ruff`
* Squashed/merged

john-parton commented:
Note: I am still missing a release note. What version of Django would this be targeting? Should be able to make it into `5.2`, I would expect.

charettes commented:
@john-parton 5.2 is feature frozen at this point so that would be 6.0.

john-parton commented:
Added a release note. Tweaked the docs to have a `versionadded` note.

I'm not the best at getting docs formatted exactly right, so some help there would be appreciated.

john-parton commented:
I should probably stop iterating, but I'm a bit obsessed at the moment, apparently.

The workaround that I had was pretty nasty in that mysql and postgres had to pass specific arguments to the method in order so that it could do the right concatenation operation.

I just pulled out the concatenation logic into its own function, and then there's actually no database-specific logic in the main workaround now.

I prepended an underscore to the function to indicate that it's private. If it's used incorrectly, it could cause pain. In particular, there's no attempt made to make sure that the expressions will actually resolve to _arrays_, and if you pass objects, the behavior is backend dependent.

john-parton commented:
Added a test that concatenating json objects preserves duplicates.

john-parton commented:
> I'd focus this PR on `JSONArray` and see what can be done for `JSONObject` in a distinct ticket/PR?

Ticket: https://code.djangoproject.com/ticket/36139
PR: https://github.com/django/django/pull/19101

john-parton commented:
Rebased, squashed, force-pushed.

john-parton commented:
Rebased, squashed, force-pushed.

@charettes Can I get some feedback on this? 

I've been using this class in production for a while and it works great on Postgres, but I do not use the other databases in production.

charettes reviewed (commented):
Hey @john-parton,

I've left pointers on how the logic can be implemented on SQLite and while clever I suggest we drop the whole array concatenation workaround for now as it would only be used on Postgres < 16 and MySQL.

john-parton commented on django/db/backends/sqlite3/features.py:
```diff
@@ -37,6 +37,8 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     supports_aggregate_order_by_clause = Database.sqlite_version_info >= (3, 44, 0)
     supports_aggregate_distinct_multiple_argument = False
     order_by_nulls_first = True
+    supports_json_absent_on_null = False
+    supports_json_array_concat = False
```
I'm not seeing the `json_array_remove_null` function anywhere in the SQLite documentation. https://www.sqlite.org/json1.html

I'm also not seeing it anywhere here: https://github.com/django/django/blob/main/django/db/backends/sqlite3/_functions.py

Could you please provide me some more direction?

john-parton replied:
I defined this function, but it should probably be noted that the `type` of the parameter is a string, not a native python list or object, so I have to use `loads`/`dumps` to modify the array.

john-parton commented on django/db/models/functions/json.py:
```diff
@@ -1,33 +1,118 @@
+from django.db.models.lookups import IsNull
+
+
+class _JSONArrayConcat(Func):
```
Workaround removed.

john-parton commented on tests/db_functions/json/test_json_concat.py:
```diff
@@ -0,0 +1,63 @@
+
+
+@skipUnlessDBFeature("supports_json_array_concat")
+class JSONArrayConcatTests(TestCase):
```
The class was just for the workaround, which was dropped, so all the tests dropped as well.

john-parton commented:
Made the requested changes, rebased, force-pushed.

I also updated the "Admonition" section of the docs to properly reflect that usage of `absent_on_null=True` is not supported on MySQL, MariaDB and PostgreSQL versions prior to 16. Let me know if any changes are required there.

It seems like a pretty bad hack to use json loads/dumps in the SQLite function. Bad code smell. Let me know if there's a more straightforward or better way of handling that.

I ran tests locally against SQLite and Postgres.

john-parton commented on django/db/models/functions/json.py:
```diff
@@ -9,29 +9,50 @@ class JSONArray(Func):
         )
 
+    def as_sqlite(self, compiler, connection, **extra_context):
+        if self.absent_on_null:
+            # SQLite does not natively support ABSENT ON NULL. Use a custom defined
+            # function to remove null values from the array.
+            extra_context["template"] = (
+                "django_json_array_remove_nulls(%(function)s(%(expressions)s))"
+            )
```
I wasn't particularly happy with this. I'm not sure the correct behavior if `template` is in extra_context already. Seems like a corner case that could possibly be ignored.

charettes reviewed (commented):
Thanks for the adjustments @john-parton, glad you figured out the SQLite function registration interface.

Left some potential pointers in how the loads/dumps roundtrip could be avoided on SQLite.

john-parton commented on django/db/backends/sqlite3/_functions.py:
```diff
@@ -300,6 +304,13 @@ def _sqlite_timestamp_diff(lhs, rhs):
     return duration_microseconds(left - right)
 
 
+def _sqlite_json_array_remove_nulls(json_array):
+    if json_array is None:
+        return None
+
+    return json.dumps([item for item in json.loads(json_array) if item is not None])
```
I like that a bit better, and we could even improve the naming of the function a tiny bit to be `django_json_array_absent_on_null` so that name is closer to the standard SQL implementation.

Additionally, the way current function is defined explodes if you put an object or other unexpected value in for `json_array`. 

I'll do some tweaks and if it passes the test cases report back.

john-parton replied:
It passed all the tests and looked good to my eye. I pushed it as an extra commit here. Let me know what you think and I can squash it down.

charettes commented on tests/db_functions/json/test_json_array.py:
```diff
@@ -107,6 +107,45 @@ def test_order_by_nested_key(self):
+            arr=JSONArray(F("name"), F("goes_by"), absent_on_null=True)
+        ).first()
+        self.assertEqual(obj.arr, ["Ivan Ivanov"])
+
```
I think it might be worth adding a test to ensure the null eliding is performed at the database level and not at the Python level.

What about

```suggestion
    @skipUnlessDBFeature("supports_json_absent_on_null")
    def test_absent_on_null_transform(self):
        obj = Author.objects.annotate(
            arr=JSONArray("goes_by", "name", absent_on_null=True),
            arr_first=F("arr__0"),
        ).first()
        self.assertEqual(obj.arr_first, "Ivan Ivanov")
```

charettes reviewed (commented):
Awesome, thanks for given the custom SQLite function a try. Kind of awesome how SQLite provides an interface for that right?

I left a suggestion for an additional test that ensures the null eliding is performed at the database level and not _faked_ at the Python level when retrieving the tuples.

Otherwise the patch LGTM 🎉 Once all squashed and rebase use the past tense in the commit message (`Add` -> `Added`) and mark the patch as not needing improvements anymore.

charettes commented on tests/db_functions/json/test_json_array.py:
```diff
@@ -107,6 +107,45 @@ def test_order_by_nested_key(self):
+
+    @skipUnlessDBFeature("supports_json_absent_on_null")
+    def test_absent_on_null(self):
+        obj = Author.objects.annotate(
+            arr=JSONArray(F("name"), F("goes_by"), absent_on_null=True)
+        ).first()
+        self.assertEqual(obj.arr, ["Ivan Ivanov"])
```
The following should ensure the null eliding takes place at the database level

```suggestion
        obj = Author.objects.annotate(
            arr=JSONArray(F("goes_by"), F("name"), absent_on_null=True),
            arr_first=F("arr__0"),
        ).first()
        self.assertEqual(obj.arr, ["Ivan Ivanov"])
        self.assertEqual(obj.arr_first, "Ivan Ivanov")
```

john-parton replied:
That's not a bad idea, and actually b y using the F expression, you're also testing that you can have the database perform operations on the array on the database side.

Should I apply the same technique to `test_single_absent_on_null` as well? I'm not entirely sure the value in even having the two tests in `test_single_absent_on_null`. Can we just drop that test?

charettes replied:
I think it might be worth keeping `test_empty_absent_on_null` as is if only to make sure there's no backend that do the crazy thing of returning `None` when there's no members at all.

john-parton replied:
Also reading over all of the JSON array tests, it's possible we might be missing something with nested JSONArray with differing absent on null.

For example, I would hope something like `JSONArray(Value(1), Value(2), JSONArray(None), absent_on_null=True)` would result in `[1, 2, [None]]` and not `[1, 2, []]` or `[1, 2]`

charettes replied:
Good catch, might be worth adding a test for nested array just to make sure.

john-parton replied:
I don't think it's necessary to test all four possible combinations of inner/outer absent_on_null, but I think just adding something equivalent to this case might be valuable

john-parton replied:
I'll get it going.

john-parton replied:
Hey, this actually revealed an error.

```
======================================================================
FAIL: test_nested_absent_on_null (db_functions.json.test_json_array.JSONArrayTests.test_nested_absent_on_null)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/john/.local/share/uv/python/cpython-3.13.0-linux-x86_64-gnu/lib/python3.13/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/home/john/.local/share/uv/python/cpython-3.13.0-linux-x86_64-gnu/lib/python3.13/unittest/case.py", line 651, in run
    self._callTestMethod(testMethod)
    
  File "/home/john/.local/share/uv/python/cpython-3.13.0-linux-x86_64-gnu/lib/python3.13/unittest/case.py", line 606, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^
  File "/home/john/Code/django/django/test/testcases.py", line 1550, in skip_wrapper
    return test_func(*args, **kwargs)
    ^^^^^^^
  File "/home/john/Code/django/tests/db_functions/json/test_json_array.py", line 162, in test_nested_absent_on_null
    self.assertEqual(obj.arr, ["Ivan Ivanov", ["iivanov", None]])
    ^^^^^^^^^^^^^^^
  File "/home/john/.local/share/uv/python/cpython-3.13.0-linux-x86_64-gnu/lib/python3.13/unittest/case.py", line 907, in assertEqual
    assertion_func(first, second, msg=msg)
    ^^^^^^^^^^^^^^^
  File "/home/john/.local/share/uv/python/cpython-3.13.0-linux-x86_64-gnu/lib/python3.13/unittest/case.py", line 1113, in assertListEqual
    self.assertSequenceEqual(list1, list2, msg, seq_type=list)
    ^^^^^^^^^^^^^^^
  File "/home/john/.local/share/uv/python/cpython-3.13.0-linux-x86_64-gnu/lib/python3.13/unittest/case.py", line 1095, in assertSequenceEqual
    self.fail(msg)
    ~~~~~~~~~^^^^^
  File "/home/john/.local/share/uv/python/cpython-3.13.0-linux-x86_64-gnu/lib/python3.13/unittest/case.py", line 732, in fail
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^
AssertionError: Lists differ: ['Ivan Ivanov', '["iivanov",null]'] != ['Ivan Ivanov', ['iivanov', None]]

First differing element 1:
'["iivanov",null]'
['iivanov', None]

- ['Ivan Ivanov', '["iivanov",null]']
?                  --       ^  ^^^ -

+ ['Ivan Ivanov', ['iivanov', None]]
?                 +        ^ +++ ^


----------------------------------------------------------------------
```

Here's the test for reference

```python
  @skipUnlessDBFeature("supports_json_absent_on_null")
  def test_nested_absent_on_null(self):
      obj = Author.objects.annotate(
          arr=JSONArray(
              F("name"),
              JSONArray(F("alias"), F("goes_by")),
              F("goes_by"),
              absent_on_null=True,
          )
      ).first()
      self.assertEqual(obj.arr, ["Ivan Ivanov", ["iivanov", None]])
```

So it looks like there's a serious issue with the custom function that it doesn't properly handle nested arrays.

charettes replied:
I suspect you won't be able to get the above test to pass on backends that don't have a native JSON type even when reverting the function. It's the case for SQLite where data is passed around as JSON serialized string.

charettes replied:
For `JSONObject` on SQLite you could use [the `json_patch('{}', json_object(...))` trick](https://sqlite.org/json1.html#jpatch) as it elides null

```sql
sqlite> SELECT json_patch('{}', json_object("dob", NULL, "name", "foo"));
{"name":"foo"}
```

john-parton commented on django/db/backends/sqlite3/_functions.py:
```diff
@@ -304,8 +304,20 @@ def _sqlite_timestamp_diff(lhs, rhs):
+    if json_array is None:
+        return None
+    deserialized = json.loads(json_array)
+    if isinstance(deserialized, dict):
+        out = {k: v for k, v in deserialized.items() if v is not None}
+    elif isinstance(deserialized, list):
+        out = [item for item in deserialized if item is not None]
```
Yeah, I was 50:50 on that.

john-parton commented:
Rebased, force-pushed. Summary of changes:

* Changed commit message to be entirely in the past tense.
* Replaced the technique of defining a `json_array_absent_on_null` SQLite function with a `json_strip_nulls` function which wraps the standard `json_array` function. Implemented only the case where `NULL` or a JSON Array is passed to the function.
* Changed my annotation. I annotated by new tests with `json_array`. Changed it to `arr` to match existing tests.
* Added a new test with nested arrays.

john-parton commented:
Does anyone know why tests failed? I'm unable to understand the error message.

Can we retrigger tests?

john-parton commented:
Amended the commit and force-pushed. Checks are going now.

john-parton commented:
We're getting `django.db.utils.OperationalError: user-defined function raised exception` on Windows SQLite, but not Linux. That's interesting. Maybe it has to do with the internal encoding of json objects? Perhaps they are bytestrings on windows instead of normal strings?

john-parton commented:
I looked again, and there's not a dedicated Linux/SQLite entry in the test matrix. It was just SQLite failing due to a mistake on my part. I've fixed it, ran the test suite locally again to 100% be sure we're good to go, and pushed again.

sarahboyce commented on docs/ref/models/database-functions.txt:
```diff
@@ -863,11 +863,42 @@ Usage example:
     >>> author.json_array
     ['margaret smith', 'msmith', 50]
 
+.. admonition:: MySQL and MariaDB
+
+    MySQL and MariaDB don't support ``True`` for the ``absent_on_null`` parameter.
+
+.. admonition:: PostgreSQL
+
+    ``True`` for the ``absent_on_null`` parameter requires PostgreSQL version 16
+    or later.
```
```suggestion
.. admonition:: MySQL, MariaDB and PostgreSQL

    The ``absent_on_null`` parameter is ignored with MySQL, MariaDB and
    PostgreSQL < 16 as they do not support it.

.. versionchanged:: 6.0

    The ``absent_on_null`` parameter was added.
```
Then remove the `.. versionadded:: 6.0` note 
We also need a release note in the 6.0 notes

john-parton replied:
Your documentation here doesn't actually reflect the behavior. It's not actually ignored, it raises a `NotSupportedError`. Is there a specific way I should phrase that?

sarahboyce replied:
```
    MySQL, MariaDB and PostgreSQL < 16 do not support the ``absent_on_null``
    parameter.
```
perhaps :thinking:

sarahboyce commented on docs/ref/models/database-functions.txt:
```diff
@@ -863,11 +863,42 @@ Usage example:
+    >>> author.json_array
+    ['margaret smith', 'msmith', 50, null]
+
+If you want to omit null values, you can do so by passing ``absent_on_null=True``:
```
```suggestion
Set the ``absent_on_null`` parameter to ``True`` to omit null values.
```
I would be tempted to combine the two usage examples so that the imports are not repeated etc. Like:

```
Accepts a list of field names or expressions and returns a JSON array
containing those values.

Set the ``absent_on_null`` parameter to ``True`` to omit null values.

Usage example:

.. code-block:: pycon

    >>> from django.db.models import F
    >>> from django.db.models.functions import JSONArray, Lower
    >>> Author.objects.create(name="Margaret Smith", alias="msmith", age=25)
    >>> author = Author.objects.annotate(
    ...     json_array=JSONArray(
    ...         Lower("name"),
    ...         "alias",
    ...         F("age") * 2,
    ...         "goes_by",
    ...     )
    ... ).get()
    >>> author.json_array
    ['margaret smith', 'msmith', 50, null]
    >>> author = Author.objects.annotate(
    ...     json_array=JSONArray(
    ...         Lower("name"),
    ...         "alias",
    ...         F("age") * 2,
    ...         "goes_by",
    ...         absent_on_null=True,
    ...     )
    ... ).get()
    >>> author.json_array
    ['margaret smith', 'msmith', 50]
```

john-parton commented:
Please be patient while I un-break my fork. I foolishly decided to try to do something on my phone.
